Update FEComponentTransfer filter effect when a child transfer function attribute changes

heycam opened the pull request:
#### 4abefb791c7642bbf2f32d92d246010370c8183d
<pre>
Update FEComponentTransfer filter effect when a child transfer function attribute changes
<a href="https://bugs.webkit.org/show_bug.cgi?id=246602">https://bugs.webkit.org/show_bug.cgi?id=246602</a>
&lt;rdar://problem/101232007&gt;

Reviewed by Said Abou-Hallawa.

Nothing currently causes the FEComponentTransfer effect to get updated
when an attribute on a child feFunc[RGBA] element changes.

Because we there are multiple child elements, we can&apos;t call
primitiveAttributeChanged on the parent SVGFEComponentTransferElement
and go through the normal setFilterEffectAttribute path like the
SVGFELightElement classes do. So instead we add a new function,
primitiveAttributeOnChildChanged, which takes the element as well.

* LayoutTests/svg/resource-invalidation/filter-resource-invalidation-expected.html:
* LayoutTests/svg/resource-invalidation/filter-resource-invalidation.html:

Add a sub-test for changing one of the feFuncX attributes.

* Source/WTF/wtf/EnumeratedArray.h:
(WTF::EnumeratedArray::index const):
(WTF::EnumeratedArray::index): Deleted.

This function needs to be const for `EnumeratedArray::operator[] const` to
work.

* Source/WebCore/platform/graphics/filters/FEComponentTransfer.cpp:
(WebCore::FEComponentTransfer::create):
(WebCore::FEComponentTransfer::FEComponentTransfer):

(WebCore::FEComponentTransfer::setType):
(WebCore::FEComponentTransfer::setSlope):
(WebCore::FEComponentTransfer::setIntercept):
(WebCore::FEComponentTransfer::setAmplitude):
(WebCore::FEComponentTransfer::setExponent):
(WebCore::FEComponentTransfer::setOffset):
(WebCore::FEComponentTransfer::setTableValues):

Setters for individual attributes of a transfer function.

(WebCore::FEComponentTransfer::externalRepresentation const):

* Source/WebCore/platform/graphics/filters/FEComponentTransfer.h:
(WebCore::FEComponentTransfer::redFunction const):
(WebCore::FEComponentTransfer::greenFunction const):
(WebCore::FEComponentTransfer::blueFunction const):
(WebCore::FEComponentTransfer::alphaFunction const):
(WebCore::FEComponentTransfer::encode const):

Store the four transfer functions in an EnumeratedArray to make access
from the new setter functions above more convenient.

* Source/WebCore/svg/SVGComponentTransferFunctionElementInlines.h: Added.

* Source/WebCore/svg/SVGComponentTransferFunctionElement.cpp:
(WebCore::SVGComponentTransferFunctionElement::svgAttributeChanged):

* Source/WebCore/svg/SVGComponentTransferFunctionElement.h:
(WebCore::SVGComponentTransferFunctionElement::channel const):
(isType):

Add type specialization support for SVGComponentTransferFunctionElement,
to avoid code that checks for each of the four feFuncX element types
explicitly.

* Source/WebCore/svg/SVGFEComponentTransferElement.cpp:
(WebCore::SVGFEComponentTransferElement::createFilterEffect const):

(WebCore::isRelevantTransferFunctionElement):

Used to ensure we only update the effect if it&apos;s the last transfer
function element of a given name that&apos;s being modified.

(WebCore::SVGFEComponentTransferElement::setFilterEffectAttributeFromChild):

Update the FEComponentTransfer effect when an attribute&apos;s changed.

(WebCore::SVGFEComponentTransferElement::transferFunctionAttributeChanged):

* Source/WebCore/svg/SVGFEComponentTransferElement.h:
* Source/WebCore/svg/SVGFilterPrimitiveStandardAttributes.cpp:
(WebCore::SVGFilterPrimitiveStandardAttributes::primitiveAttributeOnChildChanged):
* Source/WebCore/svg/SVGFilterPrimitiveStandardAttributes.h:
(WebCore::SVGFilterPrimitiveStandardAttributes::setFilterEffectAttributeFromChild):

* Source/WebCore/svg/SVGFEFuncAElement.h:
* Source/WebCore/svg/SVGFEFuncBElement.h:
* Source/WebCore/svg/SVGFEFuncGElement.h:
* Source/WebCore/svg/SVGFEFuncRElement.h:

Canonical link: <a href="https://commits.webkit.org/255869@main">https://commits.webkit.org/255869@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7c26848385e63c4cc923505d3c4b3a4a8d4c0b3e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93858 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3050 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24429 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103490 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/163824 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97850 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3066 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31285 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86178 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/99495 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99525 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2187 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80281 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/29209 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84113 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83815 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72161 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/85109 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37681 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/17642 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/80238 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35545 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18905 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27668 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4052 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39424 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/41476 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/82887 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41359 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38135 "Passed tests") | [❌ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/18727 "Found 1 new JSC stress test failure: stress/call-apply-exponential-bytecode-size.js.mini-mode (failure)") | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/svg/SVGFilterPrimitiveStandardAttributes.cpp:
```diff
@@ -155,4 +155,10 @@ void SVGFilterPrimitiveStandardAttributes::invalidateFilterPrimitiveParent(SVGEl
+void SVGFilterPrimitiveStandardAttributes::markFilterEffectForRebuildOnParent(SVGElement& element)
+{
+    if (element.parentNode() && is<SVGFilterPrimitiveStandardAttributes>(element.parentNode()))
+        downcast<SVGFilterPrimitiveStandardAttributes>(*element.parentNode()).markFilterEffectForRebuild();
```
```
if (auto* parent = dynamicDowncast<SVGFilterPrimitiveStandardAttributes>(element.parentNode()))
    parent->markFilterEffectForRebuild();
```

shallawa commented on Source/WebCore/svg/SVGComponentTransferFunctionElement.cpp:
```diff
@@ -92,7 +92,7 @@ void SVGComponentTransferFunctionElement::svgAttributeChanged(const QualifiedNam
     if (PropertyRegistry::isKnownAttribute(attrName)) {
         InstanceInvalidationGuard guard(*this);
-        SVGFilterPrimitiveStandardAttributes::invalidateFilterPrimitiveParent(this);
+        SVGFilterPrimitiveStandardAttributes::markFilterEffectForRebuildOnParent(*this);
```
It is unfortunate that we need to rebuild the whole filter effect because of changing an attribute in one of it is children. Will it be difficult to do something similar to how SVGxxxLightingElement deals with changing the light source element attributes? If it will be, I think it is okay to go this way.

heycam replied:
It's not as straightforward to do like SVGxxxLightingElement because the existing mechanism to cause updates to the effect is to go through primitiveAttributeChanged, and that function doesn't identify the element with the changed attribute (because it's assumed to be the primitive element itself, and the lighting filter primitive elements only care about a single child element, so it kind of works to call primitiveAttributeChanged).

But I guess we could add second versions of primitiveAttributeChanged / setFilterEffectAttribute that take the child element?

shallawa commented on Source/WebCore/platform/graphics/filters/FEComponentTransfer.h:
```diff
@@ -72,10 +83,7 @@ class FEComponentTransfer : public FilterEffect {
-    ComponentTransferFunction m_greenFunction;
-    ComponentTransferFunction m_blueFunction;
-    ComponentTransferFunction m_alphaFunction;
+    EnumeratedArray<Channel, ComponentTransferFunction, Channel::Alpha> m_functions;
```
Maybe

```
using Functions = EnumeratedArray<Channel, ComponentTransferFunction, Channel::Alpha>;
Functions m_functions;
```

shallawa commented on Source/WebCore/svg/SVGComponentTransferFunctionElement.h:
```diff
@@ -107,4 +109,31 @@ class SVGComponentTransferFunctionElement : public SVGElement {
+        return Channel::Red;
+    }
+}
+
```
Can not this be a pure virtual function in `SVGComponentTransferFunctionElement` and be overridden by the superclasses: `SVGFEFuncAElement`, `SVGFEFuncBElement`, etc?

heycam replied:
Yeah, I can do that. I guess it's more straightforward than computing it based on the element name.

shallawa commented on Source/WebCore/svg/SVGComponentTransferFunctionElement.h:
```diff
@@ -107,4 +109,31 @@ class SVGComponentTransferFunctionElement : public SVGElement {
+    {
+        using namespace WebCore::ElementNames::SVG;
+        auto name = element.elementName();
+        return name == feFuncR || name == feFuncG || name == feFuncB || name == feFuncA;
```
Can not this function be expressed using the superclasses traits instead:
```
return is<SVGFEFuncRElement> || is<SVGFEFuncGElement> || ...
```

heycam replied:
Oh, yes it can. That looks better.

shallawa commented on Source/WebCore/svg/SVGFEComponentTransferElement.cpp:
```diff
@@ -75,23 +76,74 @@ void SVGFEComponentTransferElement::svgAttributeChanged(const QualifiedName& att
-            alpha = downcast<SVGFEFuncAElement>(child).transferFunction();
+    using Channel = FEComponentTransfer::Channel;
+
+    EnumeratedArray<Channel, ComponentTransferFunction, Channel::Alpha> functions;
```
`FEComponentTransfer::Functions functions;`

shallawa commented on Source/WebCore/svg/SVGFEComponentTransferElement.cpp:
```diff
@@ -75,23 +76,74 @@ void SVGFEComponentTransferElement::svgAttributeChanged(const QualifiedName& att
+    return FEComponentTransfer::create(functions[Channel::Red], functions[Channel::Green], functions[Channel::Blue], functions[Channel::Alpha]);
+}
+
+static bool isRelevantTransferFunctionElement(const Element& child)
```
I am not sure what this function returns. Does it take care of the case where multiple functions of the same type are children of  SVGComponentTransferFunctionElement, for example:

           <feComponentTransfer>
                 <feFuncR type="linear" slope="1" intercept="0"/>
                 <feFuncG type="linear" slope="1" intercept="0"/>
                 <feFuncR type="linear" slope="1" intercept="1"/>
             </feComponentTransfer>

heycam replied:
Yes, exactly. The last instance of a given feFuncX child should win.

shallawa commented on Source/WebCore/platform/graphics/filters/FEComponentTransfer.h:
```diff
@@ -55,10 +56,20 @@ class FEComponentTransfer : public FilterEffect {
-    ComponentTransferFunction greenFunction() const { return m_greenFunction; }
-    ComponentTransferFunction blueFunction() const { return m_blueFunction; }
-    ComponentTransferFunction alphaFunction() const { return m_alphaFunction; }
+    enum class Channel : uint8_t { Red, Green, Blue, Alpha };
```
Maybe this can be a global enum similar to `ComponentTransferType` since it difficult to decide whether it belongs to `FEComponentTransfer`, `SVGComponentTransferFunctionElement` or `SVGFEComponentTransferElement`

heycam committed:
Update FEComponentTransfer filter effect when a child transfer function attribute changes
https://bugs.webkit.org/show_bug.cgi?id=246602
<rdar://problem/101232007>

Reviewed by Said Abou-Hallawa.

Nothing currently causes the FEComponentTransfer effect to get updated
when an attribute on a child feFunc[RGBA] element changes.

Because we there are multiple child elements, we can't call
primitiveAttributeChanged on the parent SVGFEComponentTransferElement
and go through the normal setFilterEffectAttribute path like the
SVGFELightElement classes do. So instead we add a new function,
primitiveAttributeOnChildChanged, which takes the element as well.

* LayoutTests/svg/resource-invalidation/filter-resource-invalidation-expected.html:
* LayoutTests/svg/resource-invalidation/filter-resource-invalidation.html:

Add a sub-test for changing one of the feFuncX attributes.

* Source/WTF/wtf/EnumeratedArray.h:
(WTF::EnumeratedArray::index const):
(WTF::EnumeratedArray::index): Deleted.

This function needs to be const for `EnumeratedArray::operator[] const` to
work.

* Source/WebCore/platform/graphics/filters/FEComponentTransfer.cpp:
(WebCore::FEComponentTransfer::create):
(WebCore::FEComponentTransfer::FEComponentTransfer):

(WebCore::FEComponentTransfer::setType):
(WebCore::FEComponentTransfer::setSlope):
(WebCore::FEComponentTransfer::setIntercept):
(WebCore::FEComponentTransfer::setAmplitude):
(WebCore::FEComponentTransfer::setExponent):
(WebCore::FEComponentTransfer::setOffset):
(WebCore::FEComponentTransfer::setTableValues):

Setters for individual attributes of a transfer function.

(WebCore::FEComponentTransfer::externalRepresentation const):

* Source/WebCore/platform/graphics/filters/FEComponentTransfer.h:
(WebCore::FEComponentTransfer::redFunction const):
(WebCore::FEComponentTransfer::greenFunction const):
(WebCore::FEComponentTransfer::blueFunction const):
(WebCore::FEComponentTransfer::alphaFunction const):
(WebCore::FEComponentTransfer::encode const):

Store the four transfer functions in an EnumeratedArray to make access
from the new setter functions above more convenient.

* Source/WebCore/svg/SVGComponentTransferFunctionElementInlines.h: Added.

* Source/WebCore/svg/SVGComponentTransferFunctionElement.cpp:
(WebCore::SVGComponentTransferFunctionElement::svgAttributeChanged):

* Source/WebCore/svg/SVGComponentTransferFunctionElement.h:
(WebCore::SVGComponentTransferFunctionElement::channel const):
(isType):

Add type specialization support for SVGComponentTransferFunctionElement,
to avoid code that checks for each of the four feFuncX element types
explicitly.

* Source/WebCore/svg/SVGFEComponentTransferElement.cpp:
(WebCore::SVGFEComponentTransferElement::createFilterEffect const):

(WebCore::isRelevantTransferFunctionElement):

Used to ensure we only update the effect if it's the last transfer
function element of a given name that's being modified.

(WebCore::SVGFEComponentTransferElement::setFilterEffectAttributeFromChild):

Update the FEComponentTransfer effect when an attribute's changed.

(WebCore::SVGFEComponentTransferElement::transferFunctionAttributeChanged):

* Source/WebCore/svg/SVGFEComponentTransferElement.h:
* Source/WebCore/svg/SVGFilterPrimitiveStandardAttributes.cpp:
(WebCore::SVGFilterPrimitiveStandardAttributes::primitiveAttributeOnChildChanged):
* Source/WebCore/svg/SVGFilterPrimitiveStandardAttributes.h:
(WebCore::SVGFilterPrimitiveStandardAttributes::setFilterEffectAttributeFromChild):

* Source/WebCore/svg/SVGFEFuncAElement.h:
* Source/WebCore/svg/SVGFEFuncBElement.h:
* Source/WebCore/svg/SVGFEFuncGElement.h:
* Source/WebCore/svg/SVGFEFuncRElement.h:

Canonical link: https://commits.webkit.org/255869@main
